aten/src/ATen/native/cuda/Sorting.cu

```diff
@@ -0,0 +1,451 @@
+        THCRoundUp(slice_size, (int64_t)C10_WARP_SIZE), (int64_t)1024));
+    auto stream = at::cuda::getCurrentCUDAStream();
+    gatherMedian<scalar_t, index_t, all_dims><<<grid, block, 0, stream>>>(
+        values_info,
+        indices_info,
+        self_info,
+        slice_size,
+        num_slices,
+        self_info.strides[collapse_self_dim],
+        ignore_nan);
+  }
+};
+
+template <typename scalar_t>
+void kthvalue_cuda_template(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t k,
+    int64_t dim_,
+    bool keepdim) {
+  int64_t dim = maybe_wrap_dim(dim_, self.dim());
+  int64_t slicesize = self.size(dim);
+  // FIXME: This seems bogus, I only do this because it was the old behaviour.
+  //        The reductions are fine, as long as the axis being reduced along
+  //        isn't of 0 elements (and the output has elements).
+  TORCH_CHECK(
+      self.numel() > 0,
+      "cannot perform reduction function kthvalue",
+      " on tensor with no elements because the operation does not have an identity");
+  TORCH_CHECK(k >= 1 && k <= slicesize, "selected number k out of range");
+
+  _reduction_with_indices_allocate_or_resize_output(
+      values, indices, self, dim, keepdim);
+  if (self.dim() == 0 && self.numel() == 1) {
+    values.copy_(self);
+    indices.zero_();
+    return;
+  }
+
+  TORCH_CHECK(
+      self.dim() <= MAX_TENSORINFO_DIMS,
+      "cannot operate on more than ",
+      MAX_TENSORINFO_DIMS,
+      " dimensions");
+
+  // Based on required index size, run the algorithm with the
+  // appropriate index type
+  if (cuda::detail::canUse32BitIndexMath(self) &&
+      cuda::detail::canUse32BitIndexMath(values) &&
+      cuda::detail::canUse32BitIndexMath(indices)) {
+    run_launcher<scalar_t, uint32_t>(
+        values, indices, self, dim, KthValueLauncher(k));
+  } else {
+    run_launcher<scalar_t, uint64_t>(
+        values, indices, self, dim, KthValueLauncher(k));
+  }
+
+  if (!keepdim) {
+    values.squeeze_(dim);
+    indices.squeeze_(dim);
+  }
+
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cuda(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t k,
+    int64_t dim,
+    bool keepdim) {
+  AT_DISPATCH_ALL_TYPES_AND(
+      at::ScalarType::Half, self.scalar_type(), "kthvalue_cuda", [&] {
+        kthvalue_cuda_template<scalar_t>(
+            values, indices, self, k, dim, keepdim);
+      });
+  return std::forward_as_tuple(values, indices);
+}
+
+std::tuple<Tensor&, Tensor&> median_with_indices_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim,
+    bool ignore_nan) {
+  NoNamesGuard guard;
+
+  dim = at::maybe_wrap_dim(dim, self.dim());
+  Tensor in = self.dim() > 0 ? self : self.unsqueeze(0);
+
+  int64_t size = in.size(dim);
+  TORCH_CHECK(
+      size > 0,
+      "median() operation does not have an identity for empty input tensor");
```

Comment:
This is misleading because it's not that the input tensor has no elements, but that this dimension has size zero. In NumPy the error would be something like "median of empty slice." In PyTorch we could say something like, "requested median for a dimension of size zero"?